File: src/test/java/commands/misc/BrotherCommandTest.java
package commands.misc;

import commands.CommandTest;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Mentions;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import net.dv8tion.jda.api.entities.emoji.RichCustomEmoji;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import toby.command.CommandContext;
import toby.command.commands.misc.BrotherCommand;
import toby.emote.Emotes;
import toby.jpa.dto.BrotherDto;
import toby.jpa.dto.UserDto;
import toby.jpa.service.IBrotherService;

import java.util.Optional;

import static org.mockito.Mockito.*;

class BrotherCommandTest implements CommandTest {

    @Mock
    private IBrotherService brotherService;

    @Mock
    private Guild guild;

    @Mock
    private Emoji tobyEmote;

    @Mock
    private JDA jda;

    private BrotherCommand brotherCommand;

    @BeforeEach
    void setUp() {
        setUpCommonMocks();
        brotherService = mock(IBrotherService.class);
        guild = mock(Guild.class);
        tobyEmote = mock(RichCustomEmoji.class);
        jda = mock(JDA.class);

        // Initialize the BrotherCommand with the mocked dependencies
        brotherCommand = new BrotherCommand(brotherService);
    }

    @AfterEach
    void tearDown(){
        tearDownCommonMocks();
        reset(guild);
        reset(jda);
    }

    @Test
    void testDetermineBrother_BrotherExistsWithNoMention() {
        // Arrange
        Optional<Mentions> mentions = Optional.empty();
        UserDto user = mock(UserDto.class);
        BrotherDto brotherDto = new BrotherDto();
        brotherDto.setBrotherName("TestBrother");

        OptionMapping optionMapping = mock(OptionMapping.class);
        Mockito.when(event.getOption("brother")).thenReturn(optionMapping);
        Mockito.when(event.getUser()).thenReturn(mock(User.class));
        Mockito.when(brotherService.getBrotherById(user.getDiscordId())).thenReturn(Optional.of(brotherDto));
        Mockito.when(event.getGuild()).thenReturn(guild);
        Mockito.when(guild.getJDA()).thenReturn(jda);
        Mockito.when(jda.getEmojiById(Emotes.TOBY)).thenReturn((RichCustomEmoji) tobyEmote);

        // Act
        brotherCommand.handle(new CommandContext(event), null, 0);

        // Assert
        // Verify that the expected response is sent
        verify(interactionHook, times(1)).sendMessageFormat(anyString(), eq(brotherDto.getBrotherName()));
        // You can use Mockito.verify and ArgumentMatchers to assert the behavior
    }

    @Test
    void testDetermineBrother_BrotherDoesntExistWithNoMention() {
        // Arrange
        Optional<Mentions> mentions = Optional.empty();
        UserDto userDto = mock(UserDto.class);
        User user = mock(User.class);
        BrotherDto brotherDto = new BrotherDto();
        brotherDto.setBrotherName("TestBrother");

        OptionMapping optionMapping = mock(OptionMapping.class);
        Mockito.when(event.getOption("brother")).thenReturn(optionMapping);
        Mockito.when(event.getUser()).thenReturn(user);
        Mockito.when(user.getName()).thenReturn("userName");
        Mockito.when(brotherService.getBrotherById(userDto.getDiscordId())).thenReturn(Optional.empty());
        Mockito.when(event.getGuild()).thenReturn(guild);
        Mockito.when(guild.getJDA()).thenReturn(jda);
        Mockito.when(jda.getEmojiById(Emotes.TOBY)).thenReturn((RichCustomEmoji) tobyEmote);

        // Act
        brotherCommand.handle(new CommandContext(event), null, 0);

        // Assert
        // Sends an angry message saying you're not my fucking brother with the toby emoji
        verify(interactionHook, times(1)).sendMessageFormat(anyString(), anyString(), eq(tobyEmote));
        // You can use Mockito.verify and ArgumentMatchers to assert the behavior
    }

    @Test
    void testDetermineBrother_NoBrother() {
        // Similar to the previous test but for the case where no brother is found
    }
}

File: discord-bot/src/test/kotlin/bot/toby/command/commands/misc/BrotherCommandTest.kt
package bot.toby.command.commands.misc

import bot.toby.command.CommandTest
import bot.toby.command.CommandTest.Companion.event
import bot.toby.command.CommandTest.Companion.requestingUserDto
import bot.toby.command.DefaultCommandContext
import database.dto.BrotherDto
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class BrotherCommandTest : CommandTest {
    private lateinit var brotherService: database.service.BrotherService
    private lateinit var brotherCommand: BrotherCommand

    @BeforeEach
    fun setUp() {
        setUpCommonMocks()
        brotherService = mockk(relaxed = true)
        brotherCommand = BrotherCommand(brotherService)
    }

    @AfterEach
    fun tearDown() {
        tearDownCommonMocks()
        clearMocks(brotherService)
    }

    private fun userOption(id: Long, name: String): OptionMapping {
        val u = mockk<User>(relaxed = true) {
            every { idLong } returns id
            every { effectiveName } returns name
        }
        return mockk { every { asUser } returns u }
    }

    private fun stringOption(value: String): OptionMapping =
        mockk { every { asString } returns value }

    // ---- check ----

    @Test
    fun `check returns brother message when caller is registered`() {
        every { event.subcommandName } returns BrotherCommand.CHECK
        every { event.getOption("user") } returns null
        every { brotherService.getBrotherById(1L) } returns BrotherDto(1L, "TestBrother")

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify { event.hook.sendMessage("Of course, TestBrother is one of my brothers.") }
    }

    @Test
    fun `check returns negative message when caller is not registered`() {
        every { event.subcommandName } returns BrotherCommand.CHECK
        every { event.getOption("user") } returns null
        every { brotherService.getBrotherById(any()) } returns null

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify { event.hook.sendMessage("Effective Name is not registered as a brother.") }
    }

    @Test
    fun `check uses mentioned user when one is provided`() {
        every { event.subcommandName } returns BrotherCommand.CHECK
        every { event.getOption("user") } returns userOption(42L, "OtherUser")
        every { brotherService.getBrotherById(42L) } returns BrotherDto(42L, "MentionedBrother")

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 1) { brotherService.getBrotherById(42L) }
        verify { event.hook.sendMessage("Of course, MentionedBrother is one of my brothers.") }
    }

    // ---- list ----

    @Test
    fun `list sends an embed when brothers exist`() {
        every { event.subcommandName } returns BrotherCommand.LIST
        every { brotherService.listBrothers() } returns listOf(
            BrotherDto(1L, "Alice"),
            BrotherDto(2L, "Bob"),
        )

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify { event.hook.sendMessageEmbeds(any<MessageEmbed>(), *anyVararg()) }
    }

    @Test
    fun `list reports empty state when nobody is registered`() {
        every { event.subcommandName } returns BrotherCommand.LIST
        every { brotherService.listBrothers() } returns emptyList()

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify { event.hook.sendMessage("No brothers are registered yet.") }
    }

    // ---- add ----

    @Test
    fun `add creates a brother when caller is a superuser and entry is new`() {
        every { event.subcommandName } returns BrotherCommand.ADD
        every { event.getOption("user") } returns userOption(100L, "Newbie")
        every { event.getOption("name") } returns stringOption("Newbie")
        every { requestingUserDto.superUser } returns true
        every { brotherService.getBrotherById(100L) } returns null

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 1) {
            brotherService.createNewBrother(match { it.discordId == 100L && it.brotherName == "Newbie" })
        }
        verify { event.hook.sendMessage("Registered Newbie as 'Newbie'.") }
    }

    @Test
    fun `add refuses non-superusers`() {
        every { event.subcommandName } returns BrotherCommand.ADD
        every { requestingUserDto.superUser } returns false

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 0) { brotherService.createNewBrother(any()) }
    }

    @Test
    fun `add rejects duplicates`() {
        every { event.subcommandName } returns BrotherCommand.ADD
        every { event.getOption("user") } returns userOption(100L, "Newbie")
        every { event.getOption("name") } returns stringOption("Newbie")
        every { requestingUserDto.superUser } returns true
        every { brotherService.getBrotherById(100L) } returns BrotherDto(100L, "Existing")

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 0) { brotherService.createNewBrother(any()) }
        verify { event.hook.sendMessage("Newbie is already registered as 'Existing'.") }
    }

    // ---- remove ----

    @Test
    fun `remove deletes when caller is superuser and entry exists`() {
        every { event.subcommandName } returns BrotherCommand.REMOVE
        every { event.getOption("user") } returns userOption(100L, "Goner")
        every { requestingUserDto.superUser } returns true
        every { brotherService.getBrotherById(100L) } returns BrotherDto(100L, "Goner")

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 1) { brotherService.deleteBrotherById(100L) }
        verify { event.hook.sendMessage("Unregistered Goner.") }
    }

    @Test
    fun `remove refuses non-superusers`() {
        every { event.subcommandName } returns BrotherCommand.REMOVE
        every { requestingUserDto.superUser } returns false

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 0) { brotherService.deleteBrotherById(any()) }
    }

    @Test
    fun `remove reports when entry doesn't exist`() {
        every { event.subcommandName } returns BrotherCommand.REMOVE
        every { event.getOption("user") } returns userOption(100L, "NotHere")
        every { requestingUserDto.superUser } returns true
        every { brotherService.getBrotherById(100L) } returns null

        brotherCommand.handle(DefaultCommandContext(event), requestingUserDto, 0)

        verify(exactly = 0) { brotherService.deleteBrotherById(any()) }
        verify { event.hook.sendMessage("NotHere isn't registered as a brother.") }
    }

    // ---- subcommand surface ----

    @Test
    fun `subCommands lists all four operations`() {
        val names = brotherCommand.subCommands.map { it.name }.toSet()
        assertEquals(setOf("check", "list", "add", "remove"), names)
    }
}
